storage/txn_cache.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: yananzhi (zac.zhiyanan@gmail.com)
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/cache"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+const (
+	// The default size of the txn cache.
+	defaultTxnCacheSize = 1 << 20
+)
+
+// TxnCache maintains txn which was pushed recently to avoid
+// repushing further intents after a txn has already been aborted,
+// comiitted or its timestamp moved forward.
+type TxnCache struct {
+	// Use a LRU un-ordered cache, key is txn's ID, value is txn
+	txnCache *cache.UnorderedCache
+
+	// cacheMu protects txnCache for concurrent access
+	cacheMu sync.Mutex
+}
+
+// NewTxnCache returns a new txn cache
```

Comment:
Use punctuation to end your comments (i.e. end your sentences with a period).  This is especially important for exported methods (uppercase) that will be picked up by godoc.

I would have formatted this comment like this:
`// NewTxnCache returns a new TxnCache with the given maximum size.